File: tools/count_topics.py
#!/usr/bin/env python3
import os
import sys
from collections import defaultdict

def frontmatter_categories(content):
    if not content.startswith('---'):
        return []

    frontmatter_lines = []
    past_opening = False
    closed = False

    for line in content.splitlines():
        stripped = line.rstrip('\r')
        if stripped == '---':
            if not past_opening:
                past_opening = True
                continue
            closed = True
            break
        if past_opening:
            frontmatter_lines.append(stripped)

    if not closed:
        return []

    categories = []
    in_block = False

    for line in frontmatter_lines:
        trimmed = line.strip()

        if trimmed.lower().startswith('categories:'):
            value = trimmed[len('categories:'):].strip()

            if value.startswith('[') and value.endswith(']'):
                inner = value[1:-1]
                categories = [
                    item.strip().strip('"\'')
                    for item in inner.split(',')
                    if item.strip()
                ]
                in_block = False
            elif not value:
                in_block = True
            else:
                categories = [value.strip('"\'')]
                in_block = False

        elif in_block:
            if trimmed.startswith('- '):
                val = trimmed[2:].strip().strip('"\'')
                if val:
                    categories.append(val)
            elif trimmed and not trimmed.startswith('#'):
                in_block = False

    return categories


post_dir = 'content/post'

try:
    files = os.listdir(post_dir)
except OSError:
    print(f'Error: cannot read {post_dir}', file=sys.stderr)
    sys.exit(1)

counts = defaultdict(int)

for filename in files:
    if not filename.endswith('.md'):
        continue
    path = os.path.join(post_dir, filename)
    try:
        with open(path, encoding='utf-8') as f:
            content = f.read()
    except OSError:
        continue

    for cat in frontmatter_categories(content):
        key = cat.lower().strip()
        if key:
            counts[key] += 1

for cat, count in sorted(counts.items(), key=lambda x: (-x[1], x[0])):
    print(f'{cat}: {count}')

File: tools/count_topics.swift
#!/usr/bin/env swift
import Foundation

func frontmatterCategories(in content: String) -> [String] {
    guard content.hasPrefix("---") else { return [] }

    let lines = content.components(separatedBy: "\n")
    var frontmatterLines: [String] = []
    var pastOpening = false
    var closed = false

    for line in lines {
        let stripped = line.hasSuffix("\r") ? String(line.dropLast()) : line
        if stripped == "---" {
            if !pastOpening { pastOpening = true; continue }
            closed = true; break
        }
        if pastOpening { frontmatterLines.append(stripped) }
    }

    guard closed else { return [] }

    var categories: [String] = []
    var inBlock = false

    for line in frontmatterLines {
        let trimmed = line.trimmingCharacters(in: .whitespaces)

        if trimmed.lowercased().hasPrefix("categories:") {
            let value = String(trimmed.dropFirst("categories:".count))
                .trimmingCharacters(in: .whitespaces)

            if value.hasPrefix("[") && value.hasSuffix("]") {
                let inner = String(value.dropFirst().dropLast())
                categories = inner.components(separatedBy: ",")
                    .map { $0.trimmingCharacters(in: .whitespaces)
                             .trimmingCharacters(in: CharacterSet(charactersIn: "\"'")) }
                    .filter { !$0.isEmpty }
                inBlock = false
            } else if value.isEmpty {
                inBlock = true
            } else {
                categories = [value.trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))]
                inBlock = false
            }
        } else if inBlock {
            if trimmed.hasPrefix("- ") {
                let val = String(trimmed.dropFirst(2))
                    .trimmingCharacters(in: .whitespaces)
                    .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
                if !val.isEmpty { categories.append(val) }
            } else if !trimmed.isEmpty && !trimmed.hasPrefix("#") {
                inBlock = false
            }
        }
    }

    return categories
}

let postDir = "content/post"
let fm = FileManager.default

guard let files = try? fm.contentsOfDirectory(atPath: postDir) else {
    fputs("Error: cannot read \(postDir)\n", stderr)
    exit(1)
}

var counts: [String: Int] = [:]

for filename in files where filename.hasSuffix(".md") {
    let path = (postDir as NSString).appendingPathComponent(filename)
    guard let content = try? String(contentsOfFile: path, encoding: .utf8) else { continue }

    for cat in frontmatterCategories(in: content) {
        let key = cat.lowercased().trimmingCharacters(in: .whitespaces)
        guard !key.isEmpty else { continue }
        counts[key, default: 0] += 1
    }
}

for (cat, count) in counts.sorted(by: { $0.value > $1.value || ($0.value == $1.value && $0.key < $1.key) }) {
    print("\(cat): \(count)")
}
